`TestClientDown4Sync` failed

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->

https://do.pingcap.net/jenkins/blue/organizations/jenkins/pingcap%2Ftiproxy%2Fghpr_unit_test/detail/ghpr_unit_test/34/pipeline

### 2. What did you expect to see? (Required)

No error

### 3. What did you see instead (Required)

```
--- FAIL: TestClientDown4Sync (3.83s)

    logger.go:25: 2024-01-15T10:45:32.190+0800	INFO	TestClientDown4Sync	configuring socket options	{"reuse-address": true, "reuse-port": true}

    logger.go:25: 2024-01-15T10:45:32.190+0800	INFO	TestClientDown4Sync	configuring peer listeners	{"listen-peer-urls": ["http://0.0.0.0:0/"]}

    logger.go:25: 2024-01-15T10:45:32.190+0800	INFO	TestClientDown4Sync	configuring client listeners	{"listen-client-urls": ["http://0.0.0.0:0/"]}

    logger.go:25: 2024-01-15T10:45:32.190+0800	INFO	TestClientDown4Sync	starting an etcd server	{"etcd-version": "3.5.6", "git-sha": "Not provided (use ./build instead of go build)", "go-version": "go1.21.5", "go-os": "linux", "go-arch": "amd64", "max-cpu-set": 96, "max-cpu-available": 96, "member-initialized": false, "name": "default", "data-dir": "/tmp/TestClientDown4Sync3643075183/001", "wal-dir": "", "wal-dir-dedicated": "", "member-dir": "/tmp/TestClientDown4Sync3643075183/001/member", "force-new-cluster": false, "heartbeat-interval": "100ms", "election-timeout": "1s", "initial-election-tick-advance": true, "snapshot-count": 100000, "max-wals": 5, "max-snapshots": 5, "snapshot-catchup-entries": 5000, "initial-advertise-peer-urls": ["http://localhost:2380/"], "listen-peer-urls": ["http://0.0.0.0:0/"], "advertise-client-urls": ["http://localhost:2379/"], "listen-client-urls": ["http://0.0.0.0:0/"], "listen-metrics-urls": [], "cors": ["*"], "host-whitelist": ["*"], "initial-cluster": "default=http://localhost:2380/", "initial-cluster-state": "new", "initial-cluster-token": "etcd-cluster", "quota-backend-bytes": 2147483648, "max-request-bytes": 1572864, "max-concurrent-streams": 4294967295, "pre-vote": true, "initial-corrupt-check": false, "corrupt-check-time-interval": "0s", "compact-check-time-enabled": false, "compact-check-time-interval": "1m0s", "auto-compaction-mode": "", "auto-compaction-retention": "0s", "auto-compaction-interval": "0s", "discovery-url": "", "discovery-proxy": "", "downgrade-check-interval": "5s"}

    logger.go:25: 2024-01-15T10:45:32.190+0800	WARN	TestClientDown4Sync	check file permission	{"error": "directory \"/tmp/TestClientDown4Sync3643075183/001\" exist, but the permission is \"drwxr-xr-x\". The recommended permission is \"-rwx------\" to prevent possible unprivileged access to the data"}

    logger.go:25: 2024-01-15T10:45:32.192+0800	INFO	TestClientDown4Sync	opened backend db	{"path": "/tmp/TestClientDown4Sync3643075183/001/member/snap/db", "took": "1.086438ms"}

    logger.go:25: 2024-01-15T10:45:32.194+0800	INFO	TestClientDown4Sync	starting local member	{"local-member-id": "8e9e05c52164694d", "cluster-id": "cdf818194e3a8c32"}

    logger.go:25: 2024-01-15T10:45:32.194+0800	INFO	TestClientDown4Sync.raft	8e9e05c52164694d switched to configuration voters=()

    logger.go:25: 2024-01-15T10:45:32.194+0800	INFO	TestClientDown4Sync.raft	8e9e05c52164694d became follower at term 0

    logger.go:25: 2024-01-15T10:45:32.194+0800	INFO	TestClientDown4Sync.raft	newRaft 8e9e05c52164694d [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]

    logger.go:25: 2024-01-15T10:45:32.194+0800	INFO	TestClientDown4Sync.raft	8e9e05c52164694d became follower at term 1

    logger.go:25: 2024-01-15T10:45:32.194+0800	INFO	TestClientDown4Sync.raft	8e9e05c52164694d switched to configuration voters=(10276657743932975437)

    logger.go:25: 2024-01-15T10:45:32.195+0800	WARN	TestClientDown4Sync	simple token is not cryptographically signed

    logger.go:25: 2024-01-15T10:45:32.200+0800	INFO	TestClientDown4Sync	kvstore restored	{"current-rev": 1}

    logger.go:25: 2024-01-15T10:45:32.201+0800	INFO	TestClientDown4Sync	starting etcd server	{"local-member-id": "8e9e05c52164694d", "local-server-version": "3.5.6", "cluster-version": "to_be_decided"}

    logger.go:25: 2024-01-15T10:45:32.201+0800	INFO	TestClientDown4Sync	started to purge file	{"dir": "/tmp/TestClientDown4Sync3643075183/001/member/snap", "suffix": "snap.db", "max": 5, "interval": "30s"}

    logger.go:25: 2024-01-15T10:45:32.201+0800	INFO	TestClientDown4Sync	started to purge file	{"dir": "/tmp/TestClientDown4Sync3643075183/001/member/snap", "suffix": "snap", "max": 5, "interval": "30s"}

    logger.go:25: 2024-01-15T10:45:32.202+0800	INFO	TestClientDown4Sync	started to purge file	{"dir": "/tmp/TestClientDown4Sync3643075183/001/member/wal", "suffix": "wal", "max": 5, "interval": "30s"}

    logger.go:25: 2024-01-15T10:45:32.201+0800	INFO	TestClientDown4Sync	started as single-node; fast-forwarding election ticks	{"local-member-id": "8e9e05c52164694d", "forward-ticks": 9, "forward-duration": "900ms", "election-ticks": 10, "election-timeout": "1s"}

    logger.go:25: 2024-01-15T10:45:32.202+0800	INFO	TestClientDown4Sync.raft	8e9e05c52164694d switched to configuration voters=(10276657743932975437)

    logger.go:25: 2024-01-15T10:45:32.202+0800	INFO	TestClientDown4Sync	added member	{"cluster-id": "cdf818194e3a8c32", "local-member-id": "8e9e05c52164694d", "added-peer-id": "8e9e05c52164694d", "added-peer-peer-urls": ["http://localhost:2380/"]}

    logger.go:25: 2024-01-15T10:45:32.202+0800	INFO	TestClientDown4Sync	serving peer traffic	{"address": "[::]:45287"}

    logger.go:25: 2024-01-15T10:45:32.202+0800	INFO	TestClientDown4Sync	cmux::serve	{"address": "[::]:45287"}

    logger.go:25: 2024-01-15T10:45:32.202+0800	INFO	TestClientDown4Sync	now serving peer/client/metrics	{"local-member-id": "8e9e05c52164694d", "initial-advertise-peer-urls": ["http://localhost:2380/"], "listen-peer-urls": ["http://0.0.0.0:0/"], "advertise-client-urls": ["http://localhost:2379/"], "listen-client-urls": ["http://0.0.0.0:0/"], "listen-metrics-urls": []}

    logger.go:25: 2024-01-15T10:45:32.495+0800	INFO	TestClientDown4Sync.raft	8e9e05c52164694d is starting a new election at term 1

    logger.go:25: 2024-01-15T10:45:32.495+0800	INFO	TestClientDown4Sync.raft	8e9e05c52164694d became pre-candidate at term 1

    logger.go:25: 2024-01-15T10:45:32.495+0800	INFO	TestClientDown4Sync.raft	8e9e05c52164694d received MsgPreVoteResp from 8e9e05c52164694d at term 1

    logger.go:25: 2024-01-15T10:45:32.495+0800	INFO	TestClientDown4Sync.raft	8e9e05c52164694d became candidate at term 2

    logger.go:25: 2024-01-15T10:45:32.495+0800	INFO	TestClientDown4Sync.raft	8e9e05c52164694d received MsgVoteResp from 8e9e05c52164694d at term 2

    logger.go:25: 2024-01-15T10:45:32.495+0800	INFO	TestClientDown4Sync.raft	8e9e05c52164694d became leader at term 2

    logger.go:25: 2024-01-15T10:45:32.495+0800	INFO	TestClientDown4Sync.raft	raft.node: 8e9e05c52164694d elected leader 8e9e05c52164694d at term 2

    logger.go:25: 2024-01-15T10:45:32.496+0800	INFO	TestClientDown4Sync	setting up initial cluster version using v2 API	{"cluster-version": "3.5"}

    logger.go:25: 2024-01-15T10:45:32.496+0800	INFO	TestClientDown4Sync	published local member to cluster through raft	{"local-member-id": "8e9e05c52164694d", "local-member-attributes": "{Name:default ClientURLs:[http://localhost:2379/]}", "request-path": "/0/members/8e9e05c52164694d/attributes", "cluster-id": "cdf818194e3a8c32", "publish-timeout": "7s"}

    logger.go:25: 2024-01-15T10:45:32.496+0800	INFO	TestClientDown4Sync	ready to serve client requests

    logger.go:25: 2024-01-15T10:45:32.496+0800	INFO	TestClientDown4Sync	require certificates to secure clients connections, disable TLS	{"tls": "server", "cfg": {}}

    logger.go:25: 2024-01-15T10:45:32.496+0800	INFO	TestClientDown4Sync	require certificates to secure clients connections, disable TLS	{"tls": "server", "cfg": {}}

    logger.go:25: 2024-01-15T10:45:32.496+0800	INFO	TestClientDown4Sync	no CA to verify server connections, disable TLS	{"tls": "client", "cfg": {}}

    logger.go:25: 2024-01-15T10:45:32.496+0800	INFO	TestClientDown4Sync	no CA to verify server connections, disable TLS	{"tls": "client", "cfg": {}}

    logger.go:25: 2024-01-15T10:45:32.496+0800	INFO	TestClientDown4Sync	connect ETCD servers	{"addrs": ["[::]:41989"]}

    logger.go:25: 2024-01-15T10:45:32.496+0800	INFO	TestClientDown4Sync	set initial cluster version	{"cluster-id": "cdf818194e3a8c32", "local-member-id": "8e9e05c52164694d", "cluster-version": "3.5"}

    logger.go:25: 2024-01-15T10:45:32.496+0800	INFO	TestClientDown4Sync	cluster version is updated	{"cluster-version": "3.5"}

    logger.go:25: 2024-01-15T10:45:32.497+0800	INFO	TestClientDown4Sync	connect ETCD servers	{"addrs": ["[::]:41989"]}

    logger.go:25: 2024-01-15T10:45:32.497+0800	INFO	TestClientDown4Sync	serving client traffic insecurely; this is strongly discouraged!	{"address": "[::]:41989"}

    logger.go:25: 2024-01-15T10:45:32.499+0800	INFO	TestClientDown4Sync	topology session is initialized

    info_test.go:93: 

        	Error Trace:	/home/jenkins/agent/workspace/pingcap/tiproxy/ghpr_unit_test/tiproxy/pkg/manager/infosync/info_test.go:93

        	Error:      	Condition never satisfied

        	Test:       	TestClientDown4Sync

    logger.go:25: 2024-01-15T10:45:36.012+0800	INFO	TestClientDown4Sync	closing etcd server	{"name": "default", "data-dir": "/tmp/TestClientDown4Sync3643075183/001", "advertise-peer-urls": ["http://localhost:2380/"], "advertise-client-urls": ["http://localhost:2379/"]}

    logger.go:25: 2024-01-15T10:45:36.012+0800	INFO	TestClientDown4Sync	skipped leadership transfer for single voting member cluster	{"local-member-id": "8e9e05c52164694d", "current-leader-member-id": "8e9e05c52164694d"}

    logger.go:25: 2024-01-15T10:45:36.013+0800	INFO	TestClientDown4Sync	stopping serving peer traffic	{"address": "[::]:45287"}

    logger.go:25: 2024-01-15T10:45:36.013+0800	INFO	TestClientDown4Sync	stopped serving peer traffic	{"address": "[::]:45287"}

    logger.go:25: 2024-01-15T10:45:36.013+0800	INFO	TestClientDown4Sync	closed etcd server	{"name": "default", "data-dir": "/tmp/TestClientDown4Sync3643075183/001", "advertise-peer-urls": ["http://localhost:2380/"], "advertise-client-urls": ["http://localhost:2379/"]}

FAIL

coverage: 82.5% of statements

FAIL	github.com/pingcap/tiproxy/pkg/manager/infosync	14.041s
```

### 4. What is your version? (Required)

<!-- Paste the output of tiproxy version -->
<!-- Paste the output of tiproxyctl version if related -->
<!-- Paste the output of SELECT tidb_version()  if related -->

master

--- a/lib/util/retry/retry.go
+++ b/lib/util/retry/retry.go
@@ -27,12 +27,18 @@ func NewBackOff(ctx context.Context, retryInterval time.Duration, retryCnt uint6
 }
 
 func Retry(o backoff.Operation, ctx context.Context, retryInterval time.Duration, retryCnt uint64) error {
+	if ctx.Err() != nil {
+		return ctx.Err()
+	}
 	bo := NewBackOff(ctx, retryInterval, retryCnt)
 	return backoff.Retry(o, bo)
 }
 
 func RetryNotify(o backoff.Operation, ctx context.Context, retryInterval time.Duration, retryCnt uint64,
 	notify backoff.Notify, notifyInterval uint64) error {
+	if ctx.Err() != nil {
+		return ctx.Err()
+	}
 	bo := NewBackOff(ctx, retryInterval, retryCnt)
 	var cnt uint64
 	return backoff.RetryNotify(o, bo, func(err error, duration time.Duration) {

--- a/pkg/manager/infosync/info_test.go
+++ b/pkg/manager/infosync/info_test.go
@@ -45,7 +45,7 @@ func TestTTLRefresh(t *testing.T) {
 				}
 			}
 			return satisfied
-		}, 3*time.Second, 100*time.Millisecond)
+		}, 10*time.Second, 100*time.Millisecond)
 	}
 }
 
@@ -65,7 +65,7 @@ func TestEtcdServerDown4Sync(t *testing.T) {
 			ttl = newTTL
 		}
 		return satisfied
-	}, 5*time.Second, 100*time.Millisecond)
+	}, 10*time.Second, 100*time.Millisecond)
 }
 
 // TTL and info are erased after the client shuts down normally.
@@ -75,7 +75,7 @@ func TestClientShutDown4Sync(t *testing.T) {
 	require.Eventually(t, func() bool {
 		ttl, info := ts.getTTLAndInfo(tiproxyTopologyPath)
 		return len(ttl) > 0 && len(info) > 0
-	}, 3*time.Second, 100*time.Millisecond)
+	}, 10*time.Second, 100*time.Millisecond)
 	ts.closeInfoSyncer()
 	ttl, info := ts.getTTLAndInfo(tiproxyTopologyPath)
 	require.True(t, len(ttl) == 0 && len(info) == 0)
@@ -88,12 +88,12 @@ func TestClientDown4Sync(t *testing.T) {
 	require.Eventually(t, func() bool {
 		ttl, info := ts.getTTLAndInfo(tiproxyTopologyPath)
 		return len(ttl) > 0 && len(info) > 0
-	}, 3*time.Second, 100*time.Millisecond)
+	}, 10*time.Second, 100*time.Millisecond)
 	ts.stopInfoSyncer()
 	require.Eventually(t, func() bool {
 		ttl, info := ts.getTTLAndInfo(tiproxyTopologyPath)
 		return len(ttl) == 0 && len(info) == 0
-	}, 3*time.Second, 100*time.Millisecond)
+	}, 10*time.Second, 100*time.Millisecond)
 }
 
 // Test that the result of GetTiDBTopology is right.
@@ -373,7 +373,7 @@ func (ts *etcdTestSuite) createEtcdServer(addr string) {
 			ts.t.Logf("start etcd failed, error: %s", err.Error())
 		}
 		return err == nil
-	}, 5*time.Second, 10*time.Millisecond)
+	}, 10*time.Second, 10*time.Millisecond)
 	<-etcd.Server.ReadyNotify()
 	ts.server = etcd
 }

